baseplate/experiments/__init__.py

```diff
@@ -0,0 +1,182 @@
+        pipeline unless any one of the following conditions are met:
+
+        1. bucketing_event_override is set to False.
+        2. The experiment specified by "name" explicitly disables bucketing
+           events.
+        3. We have already logged a bucketing event for the value specified by
+           experiment.bucketing_event_id(**kwargs) within the current request.
+
+        :param str name: Name of the experiment you want to run.
+        :param bool bucketing_event_override: (Optional) If set to True, will
+            always log bucketing events unless the experiment explicitly
+            disables them.  If set to False, will never send a bucketing event.
+            If set to None, no override will be applied.  Set to None by
+            default.
+        :param dict extra_event_fields: (Optional) Any extra fields you want to
+            add to the bucketing event.
+        :param kwargs:  Arguments that will be passed to experiment.variant to
+            determine bucketing, targeting, and overrides.
+
+        :rtype: :py:class:`str`
+        :return: Variant name if a variant is active, None otherwise.
+        """
+        experiment_config = self._get_config(name)
+        if not experiment_config:
+            return None
+
+        experiment = parse_experiment(experiment_config)
+        variant = experiment.variant(**kwargs)
+
+        do_log = True
+
+        if variant is None:
+            do_log = False
+
+        bucketing_id = experiment.bucketing_event_id(**kwargs)
+
+        if bucketing_id and bucketing_id in self._already_bucketed:
+            do_log = False
+
+        if bucketing_event_override is not None:
+            do_log = bool(bucketing_event_override)
+
+        do_log = do_log and experiment.should_log_bucketing()
+
+        if do_log:
+            self._log_bucketing_event(experiment, variant, extra_event_fields)
+            if bucketing_id:
+                self._already_bucketed.add(bucketing_id)
+
+        return variant
+
+    def _log_bucketing_event(self, experiment, variant, extra_event_fields):
+        if not self._event_queue:
+            return
+
+        extra_event_fields = extra_event_fields or {}
+
+        event_type = "bucket"
+        event = Event("bucketing_events", event_type)
+        for field, value in iteritems(extra_event_fields):
+            event.set_field(field, value)
+
+        event.set_field("variant", variant)
+        event.set_field("experiment_id", experiment.id)
+        event.set_field("experiment_name", experiment.name)
+        event.set_field("owner", experiment.owner)
+        try:
+            self._event_queue.put(event)
+        except EventTooLargeError:
+            logger.exception(
+                "The event payload was too large for the event queue."
+            )
+            if self._server_span:
+                self._server_span.log("error.kind", "events.too_large")
```

Comment:
hmm... i don't think that's right. it implies that the whole server span errored when it'd actually be more accurate to say some child span about publishing events was what errored. it would suck to mark an otherwise perfectly successful request as erroring because of this.